lib/logger/lib/logger/utils.ex

```diff
@@ -109,129 +124,62 @@ defmodule Logger.Utils do
-  defp collect_cc(?f, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?g, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?b, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?B, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?+, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?#, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?c, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?i, [a | args], used), do: {args, [a | used]}
-  defp collect_cc(?~, args, used), do: {args, used}
-  defp collect_cc(?n, args, used), do: {args, used}
-
-  defp collect_inspect(t, args, data, opts) do
-    data =
-      data
-      |> Inspect.Algebra.to_doc(opts)
-      |> Inspect.Algebra.format(opts.width)
-    {t, args, 'st~', [data]}
-  end
-
-  defp collect_value(current, [?* | t], [arg | args], used_format, used_args, opts, next)
-      when is_integer(arg) do
-    collect_cc(next, t, args, [?* | used_format], [arg | used_args],
-               put_value(opts, current, arg))
-  end
-
-  defp collect_value(current, [c | t], args, used_format, used_args, opts, next)
-      when is_integer(c) and c >= ?0 and c <= ?9 do
-    {t, c} = collect_value([c | t], [])
-    collect_cc(next, t, args, c ++ used_format, used_args,
-               put_value(opts, current, c |> :lists.reverse |> List.to_integer))
-  end
-
-  defp collect_value(_current, t, args, used_format, used_args, opts, next),
-    do: collect_cc(next, t, args, used_format, used_args, opts)
-
-  defp collect_value([c | t], buffer)
-    when is_integer(c) and c >= ?0 and c <= ?9,
-    do: collect_value(t, [c | buffer])
-
-  defp collect_value(other, buffer),
-    do: {other, buffer}
-
-  defp put_value(opts, key, value) do
-    if Map.has_key?(opts, key) do
-      Map.put(opts, key, value)
-    else
-      opts
-    end
+    format
+    |> :io_lib.scan_format(args)
+    |> handle_format_list(opts, [])
+  end
+
+  defp handle_format_list([], _opts, acc),
+    do: :lists.reverse(acc)
+  defp handle_format_list([spec | t], opts, acc) when is_map(spec),
+    do: handle_format_list(t, opts, [handle_format_spec(spec, opts) | acc])
+  defp handle_format_list([h | t], opts, acc),
+    do: handle_format_list(t, opts, [h | acc])
+
+  @inspected_format_spec %{
+    adjust: :right,
+    args: [],
+    control_char: ?s,
+    encoding: :unicode,
+    pad_char: ?\s,
+    precision: :none,
+    strings: true,
+    width: :none
+  }
+
+  defp handle_format_spec(%{control_char: char} = spec, opts) when char in 'wWpP',
+    do: %{@inspected_format_spec | args: [inspect_data(spec, opts)]}
+  defp handle_format_spec(spec, _opts),
+    do: spec
+
+  defp inspect_data(%{strings: false} = spec, opts) do
+    spec
+    |> Map.delete(:strings)
+    |> inspect_data(%{opts | charlists: :as_lists})
+  end
+  defp inspect_data(%{width: width} = spec, opts) do
+    spec
+    |> Map.delete(:width)
+    |> inspect_data(%{opts | width: width})
+  end
```

Comment:
I don't think we should be deleting fields on the map in a recursive fashion. It generates unnecessary garbage. You could handle both of those options on handle_format_spec, before you call `inspect_data`.

```
opts =
  case spec.strings do
    false -> %{opts | charlists: :as_lists}
    true -> opts
  end

opts = %{opts | width: spec.width}
```

Reply:
Well, I have another idea if we're going on this direction. Will PR it an if you dislike it just let me know!

Reply:
Excellent, let me know when it is ready for review then :)